website/source/docs/auth/aws-iam.html.md

```diff
@@ -0,0 +1,606 @@
+---
+layout: "docs"
+page_title: "Auth Backend: AWS-IAM"
+sidebar_current: "docs-auth-aws-iam"
+description: |-
+  The aws-iam backend allows automated authentication AWS IAM principals.
+---
+
+# Auth Backend: aws-iam
+
+The aws-iam auth backend provides a secure introduction mechanism for AWS IAM
+principals. This allows you to reduce the problem of securely introducing a
+Vault token to the problem of securely introducing AWS IAM credentials, which
+AWS has already solved in a number of use cases, such as via EC2 Instance
+Profiles and IAM roles attached to Lambda functions. It also allows you to have
+a consistent workflow between developers working on a local laptop with tools
+such as Hologram.
+
+## Comparison with AWS-EC2 Authentication Backend
+
+The AWS-IAM and AWS-EC2 authentication backends serve similar purposes. Both
```

Comment:
Sorry for being confusing here. What I meant was, if you accept the assumption (i.e., 'Role's AssumeRolePolicyDocument has "Service": "ec2.amazonaws.com" as the only Principal'), then yes, what I said is reliable. The problem is that the underlying assumption, fundamentally, is not generally reliable. It think it would be OK to let Vault admins understand and accept the risk of the unreliability of that assumption (based on their own setups and internal controls), but it's not a strong enough assumption to just bake into the product at this point, IMHO.

The reason it's unreliable is that the AssumeRolePolicyDocument is mutuable; it can be changed on the fly, and worse, in a way that's not really easy for an app like Vault to detect. A user with sufficient permissions could modify the AssumeRolePolicyDocument to allow that user to masquerade as any arbitrary EC2 instance (to `aws-iam`), get necessary credentials (valid for up to an hour), and then change the AssumeRolePolicyDocument back to something that looks innocuous, and Vault wouldn't really be able to tell. (What I confirmed with AWS is that the instance ID is visible, but they independently made sure I was aware of is precisely this scenario, that somebody with sufficient privileges could do something like this.)

Hence my comment (and request to AWS):
> It would be great if GetCallerIdentity would return who the original session issuer was, so that you could see definitively if the credentials were handed out to an EC2 instance in an IAM instane profile (and no longer had to rely on "Service": "ec2.amazonaws.com" being the only Principal in the AssumeRolePolicyDocument as I mentioned above).

Does this make a little more sense?

That all being said, just to add a little perspective, a user with sufficient privileges could also just launch an instance in an arbitrary subnet/VPC with arbitrary tags from an arbitrary AMI and then swap out EBS volumes to get full control of an EC2 instance in almost any conditions we might want to filter on, bypassing pretty much all of the filtering. Yes, this is a more sophisticated attack than tampering with the AssumeRolePolicyDocument to masquerade as an EC2 instance, but ultimately, the gap we're talking about here, while important, is just one of degree, not of the class of attack that pretty much any practical AWS-based authentication system I can think of would be susceptible to.